akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala

```diff
@@ -40,21 +41,31 @@ private[http] abstract class HttpMessageParser[Output >: MessageOutput <: Parser
     if (settings.includeTlsSessionInfoHeader && tlsSessionInfoHeader != null) ListBuffer(tlsSessionInfoHeader)
     else ListBuffer()
 
-  def isTerminated = terminated
-
-  val stage: PushPullStage[SessionBytes, Output] =
-    new PushPullStage[SessionBytes, Output] {
-      def onPush(input: SessionBytes, ctx: Context[Output]) = handleParserOutput(self.parseSessionBytes(input), ctx)
-      def onPull(ctx: Context[Output]) = handleParserOutput(self.onPull(), ctx)
-      private def handleParserOutput(output: Output, ctx: Context[Output]): SyncDirective =
-        output match {
-          case StreamEnd    ⇒ ctx.finish()
-          case NeedMoreData ⇒ ctx.pull()
-          case x            ⇒ ctx.push(x)
+  val stage = new GraphStage[FlowShape[SessionBytes, Output]] {
+    val in: Inlet[SessionBytes] = Inlet("HttpMessageParser.in")
+    val out: Outlet[Output] = Outlet("HttpMessageParser.out")
+    override val shape: FlowShape[SessionBytes, Output] = FlowShape(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        override def onPush(): Unit = handleParserOutput(self.parseSessionBytes(grab(in)))
+        override def onPull(): Unit = handleParserOutput(self.onPull())
```

Comment:
The caller site is;

`    Flow[SessionBytes].via(
      // each connection uses a single (private) request parser instance for all its requests
      // which builds a cache of all header instances seen on that connection
      rootParser.createShallowCopy().stage).named("rootParser")
      .map(establishAbsoluteUri)`

So it looks ok.  In any case the refactor hasn't changed the inner behaviour so it would have been a problem before right?